README
B2TAG(1)                         User Commands                        B2TAG(1)

NAME
       b2tag - detect data corruption with checksums in extended attributes.

SYNOPSIS
       b2tag [OPTION]... <FILE>...

DESCRIPTION
       b2tag is a re-implementation (in C) of shatag(1), which helps detect
       silent data corruption using checksums and extended attributes.

       b2tag writes the last modification time (mtime) and the checksum of a
       file into the file's extended attributes.

       When run again, b2tag compares stored mtime and checksum to the file's
       current mtime and checksum.  It will print the name of each file
       processed along with its status.

       File statuses are:

            backdated   checksum is wrong and mtime is older than the sha tag

            corrupt     checksum is wrong and mtime matches the sha tag

            invalid     sha tag attributes are corrupt

            new         sha tag attributes have not been set

            ok          checksum is correct and the mtime matches

            hash ok     checksum is correct and the mtime has changed

            outdated    checksum is wrong and mtime is newer than the sha tag
                        (the file has been updated since the last run)

       b2tag will automatically store the checksums and timestamps for new
       files and will update the stored checksums and timestamps for hash ok
       and outdated files (unless --dry-run is specified).

       b2tag will not update the stored checksums and timestamps for
       backdated, corrupt, or invalid files unless --force is specified.

       b2tag aims to be format-compatible with shatag and uses the same
       extended attributes (see the COMPATIBILITY section).

NOTES
       Since b2tag operates using extended attributes, any files specified
       must be stored on a filesystem that supports extended user attributes.
       Check your filesystem's documentation for more information.

       For ext2/3/4, you may need to mount the filesystem with the user_xattr
       option).

OPTIONS
   Positional Arguments
       FILE   Files to check and hash.

   Optional Arguments
       -c, --check
              Check the hashes on all specified files. Without this option,
              b2tag assumes all files with the same timestamp are OK without
              checking their contents.

       -f, --force
              Update the stored hashes for backdated, corrupted, or invalid
              files.

       -h, --help
              Output a usage message and exit.

       -n, --dry-run
              Don't create or update any extended attributes (no on-disk
              changes).  This will still read and hash the specified files.

       -p, --print
              Print the hashes of all specified files in the sha*sum format.

       -q, --quiet
              Only print errors including checksum failures. Can be specified
              multiple times to print even fewer messages. This is the
              opposite of --verbose.

       -r, --recursive
              Process directories and their contents (not just files).

       -v, --verbose
              Print more verbose messages error and warnings messages. Can be
              specified multiple times to print even more messages. This is
              the opposite of --quiet.

       -V, --version
              Output version information about b2tag and exit.

   Hash Algorithms
       --blake2b
              Use the Blake2b 512-bit hash algorithm (default). On 64-bit
              machines, blake2b is usually the fastest.

       --blake2s
              Use the Blake2s 256-bit hash algorithm. On 32-bit machines,
              blake2s is usually the fastest.

       --sha512
              Use the SHA-512 hash algorithm. On 64-bit machines, sha-512 is
              usually much faster than sha256.

       --sha256
              Use the SHA-256 hash algorithm. SHA-256 is the only hash
              algorithm supported by the original shatag utility. It is
              usually the slowest hash algorithm on 64-bit machines.

       --sha1 Use the SHA-1 hash algorithm.  SHA-1 is not secure and is not
              recommended.

       --md5  Use the MD5 hash algorithm.  MD5 is not secure and is not
              recommended.

EXAMPLES
       Create or update the stored hashes for a file:

       b2tag example.txt

       Verify that a file matches its stored attributes with the sha256 hash
       algorithm:

       b2tag -c --sha256 example.txt

       Verify all files in a directory, and filter-out OK file messages:

       b2tag -cr /example/ | grep -v ': OK$'

       Print a file's stored sha512 hashes:

       b2tag -p --sha512 example.txt > example.sha512

       This allows you to verify the file's contents with the sha512sum
       utility (e.g. after copying the files to another machine without shatag
       installed):

       sha512sum -c example.sha512

       Copy a file to a different machine and verify its contents:

       rsync -tX example.txt user@example.com:example.txt
       ssh user@example.com b2tag -cn example.txt

       Create and update the hashes for all files over 1 MB on the root
       filesystem:

       find / -xdev -type f -size +1M -print0 | xargs -r0 b2tag > b2tag.log

       To remove the extended attributes from all files:

       find / -xdev -type f -exec setfattr -x user.shatag.ts {} \; -exec
       setfattr -x user.shatag.sha256 {} \;

EXIT STATUS
       0 Success
       >0 An error occurred or at least 1 file is backdated, corrupt, or
       invalid

COMPATIBILITY
       The original shatag utility only supports the sha256 hash algorithm.

       Additionally, b2tag writes the user.shatag.ts field with full
       nanosecond precision, while python uses a floating point number for the
       whole mtime. As a result, the original shatag utility is only accurate
       to within about 200 ns. Because of this, b2tag treats small timestamps
       (fewer than 9 fractional digits) within 1 µs as equal. Timestamps with
       full nanosecond precision are compared normally.

AUTHOR
       Written by Jakob Unterwurzacher, and Tim Schlueter.

COPYRIGHT
       Copyright © 2012 Jakob Unterwurzacher.
       Copyright © 2018 Tim Schlueter.

       License: GPLv2+ with OpenSSL exception

       GNU GPL version 2 or later
       <https://www.gnu.org/licenses/licenses.html>.

       This is free software: you are free to change and redistribute it.
       There is NO WARRANTY, to the extent permitted by law.

       As a special exception to the GPL, you may link the code with the
       OpenSSL library (see the source files for more details).

SEE ALSO
       shatag(1), b2sum(1), sha256sum(1), getfattr(1), setfattr(1)

       The original shatag utility, written in python by Maxime Augier:
       https://bitbucket.org/maugier/shatag

AVAILABILITY
       Up-to-date sources can be found at:
       https://github.com/modelrockettier/b2tag

b2tag 0.2                       September 2018                        B2TAG(1)


debian/changelog
b2tag (0.2-1) unstable; urgency=medium

  * Added getopt arg parsing.
  * Added several other hash algorithms.
  * Change the hash algorithm at runtime from the command-line.
  * Hash multiple files in a single invocation.
  * Recursively hash files in a directory.
  * Print the stored file hashes in the coreutils *sum format.
  * Made blake2b the default hash algorithm.
  * Renamed utility to b2tag.

 -- Tim Schlueter <schlueter.tim@linux.com>  Sun, 23 Sep 2018 20:21:26 -0700

cshatag (0.1-1) unstable; urgency=low

  * Initial release.

 -- Tim Schlueter <schlueter.tim@linux.com>  Sun, 16 Sep 2018 12:36:44 -0700


b2tag.c
/*
 * Copyright (C) 2012 Jakob Unterwurzacher <jakobunt@gmail.com>
 * Copyright (C) 2018 Tim Schlueter
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * In addition, as a special exception, the author of this program
 * gives permission to link the code portions of this program with the
 * OpenSSL library under certain conditions as described in each file,
 * and distribute linked combinations including the two.
 * You must obey the GNU General Public License in all respects for all
 * of the code used other than OpenSSL.  If you modify this file(s)
 * with this exception, you may extend this exception to your version
 * of the file(s), but you are not obligated to do so.  If you do not
 * wish to do so, delete this exception statement from your version.
 * If you delete this exception statement from all source files in the
 * program, then also delete it here.
 */

/** @file
 * Main file for the b2tag utility (mostly command-line processing).
 */

/**
 * @mainpage
 * @include README
 */

#include "b2tag.h"

#include <assert.h>
#include <ctype.h>
#include <getopt.h>
#include <string.h>
#include <stdlib.h>

#include "file.h"
#include "utilities.h"


/** The options set by command-line arguments. */
struct args_s args;

/**
 * Prints version information for b2tag.
 *
 * @note VERSION_STRING is generated in the Makefile and defined on the compiler command-line.
 */
static void version(void)
{
	printf("b2tag version %s\n", VERSION_STRING);
}

/**
 * Prints a usage message for b2tag.
 *
 * @param program  The name of the program being run.
 */
static void usage(const char *program)
{
	printf(
		"Usage: %s [OPTION]... <FILE>...\n"
		"\n"
		"Display and update xattr-based checksums.\n"
		"\n"
		"Positional arguments:\n"
		"  FILE                  files to checksum\n"
		"\n"
		"Optional arguments:\n"
		"  -c, --check           check the hashes on all specified files\n"
		"  -f, --force           update the stored hashes for backdated, corrupted, or\n"
		"                        invalid files\n"
		"  -h, --help            show this help message and exit\n"
		"  -n, --dry-run         don't update any stored attributes\n"
		"  -p, --print           print the hashes of all specified files\n"
		"  -q, --quiet           only print errors (including checksum failures)\n"
		"  -r, --recursive       process directories and their contents (not just files)\n"
		"  -v, --verbose         print all checksums (not just missing/changed)\n"
		"  -V, --version         output version information and exit\n"
		"\n"
		"Hash algorithms:\n"
		"  --blake2b (default, 512-bit)  --blake2s (256-bit, recommended on 32-bit)\n"
		"  --sha512                      --sha256 (shatag compatible)\n"
		"  --sha1 (deprecated)           --md5 (deprecated)\n",
		program);
}

/**
 * Long options to pass to getopt.
 */
static const struct option long_opts[] = {
	{ "check",      no_argument, 0, 'c' },
	{ "dry-run",    no_argument, 0, 'n' },
	{ "force",      no_argument, 0, 'f' },
	{ "help",       no_argument, 0, 'h' },
	{ "print",      no_argument, 0, 'p' },
	{ "quiet",      no_argument, 0, 'q' },
	{ "recursive",  no_argument, 0, 'r' },
	{ "verbose",    no_argument, 0, 'v' },
	{ "version",    no_argument, 0, 'V' },
	{ "md5",        no_argument, 0,  0  },
	{ "sha1",       no_argument, 0,  0  },
	{ "sha256",     no_argument, 0,  0  },
	{ "sha512",     no_argument, 0,  0  },
	{ "blake2",     no_argument, 0,  1  },
	{ "blake2b",    no_argument, 0,  1  },
	{ "blake2b512", no_argument, 0,  1  },
	{ "blake2s",    no_argument, 0,  2  },
	{ "blake2s256", no_argument, 0,  2  },
	{ NULL, 0, 0, 0 }
};

/**
 * The entry point to the b2tag utility.
 *
 * @param argc  The number of command-line arguments.
 * @param argv  The command-line arguments.
 *
 * @retval 0  Program completed successfully.
 * @retval !0 An error occurred.
 */
int main(int argc, char *argv[])
{
	int ret = 0;
	char *program = basename(argv[0]);
	int opt;
	int option_index = 0;

	args.alg = HASH_ALG_BLAKE2B;

	while ((opt = getopt_long(argc, argv, "cfhnpqrvV", long_opts, &option_index)) != -1) {
		switch (opt) {
		case 0:
			ret = get_alg_by_name(long_opts[option_index].name, &args.alg);
			assert(ret == 0);
			break;
		case 1:
			args.alg = HASH_ALG_BLAKE2B;
			break;
		case 2:
			args.alg = HASH_ALG_BLAKE2S;
			break;
		case 'c':
			args.check = true;
			break;
		case 'f':
			args.force = true;
			break;
		case 'h':
			usage(program);
			return EXIT_SUCCESS;
		case 'n':
			args.dry_run = true;
			break;
		case 'p':
			args.print = true;
			break;
		case 'q':
			args.verbose--;
			break;
		case 'r':
			args.recursive = true;
			break;
		case 'v':
			args.verbose++;
			break;
		case 'V':
			version();
			return EXIT_SUCCESS;

		default:
			if (isprint(opt))
				fprintf(stderr, "Unknown argument '-%c' (%#02x)\n", opt, opt);
			else
				fprintf(stderr, "Unknown argument -%#02x\n", opt);

			/* Fall through. */

		case '?':
			usage(program);
			return EXIT_FAILURE;
		}
	}

	ret = 0;

	argc -= optind;
	argv += optind;

	if (argc < 1) {
		fprintf(stderr, "No file specified.\n");
		usage(program);
		return EXIT_FAILURE;
	}

	if (args.dry_run && args.force)
		pr_warn("Warning: --dry-run takes precedence over --force.\n");

	while (argc >= 1) {
		int err;
		char *pos = argv[0] + strlen(argv[0]) - 1;

		/* Remove trailing slashes */
		while (pos > argv[0] && *pos == '/')
			*pos-- = '\0';

		err = process_path(argv[0]);

		if (err < 0)
			return ret;
		else if (ret == 0 && err > 0)
			ret = err;

		argc--;
		argv++;
	}

	return ret;
}
